pkg/image/cdi/upload.go

```diff
@@ -1,20 +1,334 @@
+	}
+
+	// create DataVolume
+	dvName := cu.vmio.GetName(vmImg)
+	dvNamespace := cu.vmio.GetNamespace(vmImg)
+
+	logrus.Infof("The VM Image Status updated, start to create DataVolume %s/%s", dvNamespace, dvName)
+
+	// generate DV source
+	dvSource, err := generateDVSource(vmImg, cu.vmio.GetSourceType(vmImg))
+	if err != nil {
+		return fmt.Errorf("failed to generate DV source: %v", err)
+	}
+
+	// generate DV target storage
+	dvTargetStorage, err := generateDVTargetStorage(vmImg)
+	if err != nil {
+		return fmt.Errorf("failed to generate DV target storage: %v", err)
+	}
+	var boolTrue = true
+	dataVolumeTemplate := &cdiv1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dvName,
+			Namespace: dvNamespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         common.HarvesterAPIV1Beta1,
+					Kind:               common.VMImageKind,
+					Name:               cu.vmio.GetName(vmImg),
+					UID:                cu.vmio.GetUID(vmImg),
+					BlockOwnerDeletion: &boolTrue,
+				},
+			},
+		},
+		Spec: cdiv1.DataVolumeSpec{
+			Source:  dvSource,
+			Storage: dvTargetStorage,
+		},
+	}
+	if _, err := cu.dataVolumeClient.Create(dataVolumeTemplate); err != nil {
+		return fmt.Errorf("failed to create DataVolume %s/%s: %v", dvNamespace, dvName, err)
+	}
+	logrus.Infof("DataVolume %s/%s created", dvNamespace, dvName)
+
+	// wait data volume UploadReady
+	if err := wait.PollUntilContextTimeout(context.Background(), tickPolling, tickTimeout, true, func(context.Context) (bool, error) {
+		return cu.waitDataVolumeStatus(dvNamespace, dvName, cdiv1.UploadReady)
+	}); err != nil {
+		return fmt.Errorf("failed to wait for VMImage status: %v", err)
+	}
+
+	uploadTokenRequest := &uploadcdiv1.UploadTokenRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "token-for-upload",
+			Namespace: dvNamespace,
+		},
+		Spec: uploadcdiv1.UploadTokenRequestSpec{
+			PvcName: dvName,
+		},
+	}
+	retUploadTokenRequest, err := cu.cdiUploadClient.Create(uploadTokenRequest)
+	if err != nil {
+		return fmt.Errorf("failed to create UploadTokenRequest %s/%s: %v", dvNamespace, dvName, err)
+	}
+
+	newBody := io.MultiReader(bytes.NewReader(dataContent), req.Body)
+
+	progress := &ProgressUpdater{
+		targetBytes:       fileSize,
+		lastTime:          time.Now(),
+		vmImgUpdateLocker: updaterLocker,
+		vmImgCond:         updaterCond,
+	}
+	hookedReader := io.TeeReader(newBody, progress)
+
+	token := retUploadTokenRequest.Status.Token
+	cdiUploadURL := fmt.Sprintf("https://%s%s", CDIUploadURLRaw, UploadProxyURI)
+	uploadReq, err := http.NewRequestWithContext(req.Context(), http.MethodPost, cdiUploadURL, io.NopCloser(hookedReader))
+	if err != nil {
+		return fmt.Errorf("failed to wrap the upload request: %w", err)
+	}
+	uploadReq.Header = req.Header
+	uploadReq.Header.Add("Authorization", "Bearer "+token)
+
+	// create VMI progress updater
+	go cu.updateVMImageProgress(vmImg, updaterCond, progress, fileSize)
```

Comment:
Although we start the goroutine here, it will remain in the waiting state until the call to `Signal()` at line 261, right? 

1. If any errors happened before line 261, won't this goroutine remain stuck indefinitely?
1. Instead of locking and waiting for `Signal()` inside the goroutine, can we just let the `for` loop iterates in the background with N seconds interval?

Reply:
Thanks for the catching.
I added the defer func to handle the early error case.

BTW, we have another Single on the io writer (to trigger progress update every 1 seconds)